### PR TITLE
[Feature] [1.2.0] Linkis jdbc engine supports multiple data source links.

### PR DESCRIPTION
**Pull request by CCweixiao**
https://github.com/apache/incubator-linkis/issues/2092
Linkis jdbc engine supports multiple data source links.

image

image

### CODE REVIEW

File: linkis-engineconn-plugins/engineconn-plugins/jdbc/src/main/java/org/apache/linkis/manager/engineplugin/jdbc/JDBCDataSourceConfigurations.java [hunk 1]
```diff
@@ -43,7 +43,10 @@ public void saveStatement(String taskId, Statement statement) {
     }
 
     public void cancelStatement(String taskId) throws SQLException {
-        taskIdStatementMap.get(taskId).cancel();
+        Statement statement = taskIdStatementMap.get(taskId);
```
Comment (peacewong) [inline linkis-engineconn-plugins/engineconn-plugins/jdbc/src/main/java/org/apache/linkis/manager/engineplugin/jdbc/JDBCDataSourceConfigurations.java:46]: need to add log

File: linkis-engineconn-plugins/engineconn-plugins/jdbc/src/main/scala/org/apache/linkis/manager/engineplugin/jdbc/executer/JDBCEngineConnExecutor.scala [hunk 2]
```diff
@@ -71,10 +72,18 @@ class JDBCEngineConnExecutor(override val outputPrintLimit: Int, val id: Int) ex
 
     if (properties.get(JDBCEngineConnConstant.JDBC_URL) == null) {
       info(s"The jdbc url is empty, adding now...")
-      val globalConfig = Utils.tryAndWarn(JDBCEngineConfig.getCacheMap(engineExecutorContext.getLabels))
+      val globalConfig: util.Map[String, String] = Utils.tryAndWarn(JDBCEngineConfig.getCacheMap(engineExecutorContext.getLabels))
+      var dataSourceInfo: util.Map[String, String] = new util.HashMap[String, String]()
```
Comment (peacewong) [inline linkis-engineconn-plugins/engineconn-plugins/jdbc/src/main/scala/org/apache/linkis/manager/engineplugin/jdbc/executer/JDBCEngineConnExecutor.scala:76]: val is recommended

File: linkis-engineconn-plugins/engineconn-plugins/jdbc/src/main/scala/org/apache/linkis/manager/engineplugin/jdbc/executer/JDBCEngineConnExecutor.scala [hunk 3]
```diff
@@ -71,10 +72,18 @@ class JDBCEngineConnExecutor(override val outputPrintLimit: Int, val id: Int) ex
...
       if (StringUtils.isNotBlank(dataSourceName)) {
         info("Start getting data source connection parameters from the data source hub.")
-        // todo get data source info by data source client
+        Utils.tryCatch {
+          dataSourceInfo = JDBCMultiDatasourceParser.queryDatasourceInfoByName(dataSourceName, execSqlUser)
```
Comment (peacewong) [inline linkis-engineconn-plugins/engineconn-plugins/jdbc/src/main/scala/org/apache/linkis/manager/engineplugin/jdbc/executer/JDBCEngineConnExecutor.scala:80]: Judge NPE here

File: linkis-engineconn-plugins/engineconn-plugins/jdbc/src/main/scala/org/apache/linkis/manager/engineplugin/jdbc/executer/JDBCEngineConnExecutor.scala [hunk 4]
```diff
@@ -71,10 +72,18 @@ class JDBCEngineConnExecutor(override val outputPrintLimit: Int, val id: Int) ex
...
+        } {
+          e: Throwable => return ErrorExecuteResponse(e.getMessage, e)
```
Comment (peacewong) [inline linkis-engineconn-plugins/engineconn-plugins/jdbc/src/main/scala/org/apache/linkis/manager/engineplugin/jdbc/executer/JDBCEngineConnExecutor.scala:82]: It is suggested to modify the message to “Failed to get datasource info from datasource server”

**Comment by CCweixiao** [comment]
平哥，多数据的有一个问题，就是数据源权限，可见性的问题，默认用户只能看到自己加的数据源信息，如果对于同一个数据源？每个用户都全部加一次？感觉这样不太好。数据源应该由管理员统一维护，然后每个用户去申请自己需要的数据源，管理员审批许可就行

File: linkis-engineconn-plugins/engineconn-plugins/jdbc/src/main/scala/org/apache/linkis/manager/engineplugin/jdbc/executer/JDBCMultiDatasourceParser.scala [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+object JDBCMultiDatasourceParser extends Logging {
+
+  def queryDatasourceInfoByName(datasourceName: String, username: String): util.Map[String, String] = {
+    info(s"Starting query [$username, $datasourceName] datasource info ......")
+    val dataSourceClient = new LinkisDataSourceRemoteClient()
+    var dataSource: DataSource = null
+    Utils.tryCatch {
+      dataSource = dataSourceClient.getInfoByDataSourceName(GetInfoByDataSourceNameAction.builder()
+        .setSystem("")
```
Comment (peacewong) [inline linkis-engineconn-plugins/engineconn-plugins/jdbc/src/main/scala/org/apache/linkis/manager/engineplugin/jdbc/executer/JDBCMultiDatasourceParser.scala:39]: It is recommended that the parameter be passed in, the default is empty

File: linkis-engineconn-plugins/engineconn-plugins/jdbc/src/main/scala/org/apache/linkis/manager/engineplugin/jdbc/executer/JDBCMultiDatasourceParser.scala [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+      throw JDBCParamsIllegalException("The data source jdbc connection port cannot be null!")
+    }
+    var jdbcUrl = s"jdbc:$dbType://$host:$port"
+    val dbName = dbConnParams.get(JDBCEngineConnConstant.DS_JDBC_DB_NAME)
+    if (dbName != null && StringUtils.isNotBlank(dbName.toString)) {
+      jdbcUrl = s"$jdbcUrl/$dbName"
+    }
+
+    val params = dbConnParams.get(JDBCEngineConnConstant.DS_JDBC_PARAMS)
+    var paramsMap: util.Map[String, Object] = new util.HashMap[String, Object]()
```
Comment (peacewong) [inline linkis-engineconn-plugins/engineconn-plugins/jdbc/src/main/scala/org/apache/linkis/manager/engineplugin/jdbc/executer/JDBCMultiDatasourceParser.scala:102]: val is recommended

**Comment by peacewong** [comment]
> 平哥，多数据的有一个问题，就是数据源权限，可见性的问题，默认用户只能看到自己加的数据源信息，如果对于同一个数据源？每个用户都全部加一次？感觉这样不太好。数据源应该由管理员统一维护，然后每个用户去申请自己需要的数据源，管理员审批许可就行
> 
> Brother Ping, there is a problem with multiple data, that is, the problem of data source permissions and exposure. By default, users can only see the data source information added by themselves. If it is the same data source? Add all users once? The data should be maintained uniformly by the administrator, and then each source should apply for the data source it needs, and the administrator should review it.

This is a must, excellent feature

**Comment by peacewong** [comment]
> 平哥，多数据的有一个问题，就是数据源权限，可见性的问题，默认用户只能看到自己加的数据源信息，如果对于同一个数据源？每个用户都全部加一次？感觉这样不太好。数据源应该由管理员统一维护，然后每个用户去申请自己需要的数据源，管理员审批许可就行
> 
> Brother Ping, there is a problem with multiple data, that is, the problem of data source permissions and exposure. By default, users can only see the data source information added by themselves. If it is the same data source? Add all users once? The data should be maintained uniformly by the administrator, and then each source should apply for the data source it needs, and the administrator should review it.

This is a must, excellent feature

**Review by peacewong — APPROVED** [review]
LGTM.

**Pull request merged**